### PR TITLE
Add remaining tests of CEReactions for the DOM specification.

### PR DESCRIPTION
**Pull request by rniwa**
Now that we have the basic testing infrastructure for CEReactions, add tests for the remaining interfaces in https://dom.spec.whatwg.org/

Most of tests pass on Chrome, and we intend to fix the behavior in WebKit.

### CODE REVIEW

File: custom-elements/reactions/DOMTokenList.html [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+<title>Custom Elements: CEReactions on DOMTokenList interface</title>
+<meta name="author" title="Ryosuke Niwa" href="mailto:rniwa@webkit.org">
+<meta name="assert" content="add, remove, toggle, replace, and the stringifier of DOMTokenList interface must have CEReactions">
+<meta name="help" content="https://dom.spec.whatwg.org/#node">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="./resources/reactions.js"></script>
+</head>
+<body>
+<div id="log"></div>
```
Comment (dominiccooney) [inline custom-elements/reactions/DOMTokenList.html]: Is this div necessary?

File: custom-elements/reactions/DOMTokenList.html [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    var instance = document.createElement(element.name);
+    assert_array_equals(element.log().types(), ['constructed']);
+    instance.classList.add('foo');
+    var logEntries = element.log();
+    assert_array_equals(logEntries.types(), ['attributeChanged']);
+    assert_equals(logEntries.log().name, 'class');
+    assert_equals(logEntries.log().oldValue, null);
+    assert_equals(logEntries.log().newValue, 'foo');
+    assert_equals(logEntries.log().namespace, null);
+}, 'add on DOMTokenList must enqueue a attributeChanged reaction when adding an attribute');
```
Comment (dominiccooney) [inline custom-elements/reactions/DOMTokenList.html]: Minor grammar point, but maybe "a attributeChanged" &rarr; "an attributeChanged"; below too

File: custom-elements/reactions/DOMTokenList.html [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+}, 'add on DOMTokenList must enqueue a attributeChanged reaction when adding a value to an existing attribute');
+
+test(function () {
+    var element = defineNewCustomElement([]);
+    var instance = document.createElement(element.name);
+    instance.setAttribute('class', 'hello');
+    assert_array_equals(element.log().types(), ['constructed']);
+    instance.classList.add('world');
+    assert_array_equals(element.log().types(), []);
+}, 'add on DOMTokenList must enqueue a attributeChanged reaction when adding a value to an existing unobserved attribute');
```
Comment (dominiccooney) [inline custom-elements/reactions/DOMTokenList.html]: It must *not* enqueue an attributeChanged reaction. The test looks right but the description is confusing.

File: custom-elements/reactions/DOMTokenList.html [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+}, 'remove on DOMTokenList must not enqueue a attributeChanged reaction when removing a non-existent value from an attribute');
+
+test(function () {
+    var element = defineNewCustomElement([]);
+    var instance = document.createElement(element.name);
+    instance.setAttribute('class', 'hello world');
+    assert_array_equals(element.log().types(), ['constructed']);
+    instance.classList.remove('world');
+    assert_array_equals(element.log().types(), []);
+}, 'remove on DOMTokenList not must enqueue a attributeChanged reaction when removing a value from an unobserved attribute');
```
Comment (dominiccooney) [inline custom-elements/reactions/DOMTokenList.html]: not must &rarr; must not

There's one other instance of this below, too.

File: custom-elements/reactions/DOMTokenList.html [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    instance.classList = 'hello';
+    var logEntries = element.log();
+    assert_array_equals(logEntries.types(), ['attributeChanged']);
+    assert_equals(logEntries.log().name, 'class');
+    assert_equals(logEntries.log().oldValue, 'hello');
+    assert_equals(logEntries.log().newValue, 'hello');
+    assert_equals(logEntries.log().namespace, null);
+}, 'the stringifier of DOMTokenList must enqueue a attributeChanged reaction when the setter is called with the original value of the attribute');
+
+</script>
```
Comment (dominiccooney) [inline custom-elements/reactions/DOMTokenList.html]: I wonder if we also need tests for eg add('foo', 'bar') that check we only run "update steps" once and hence only one attributeChanged callback?

Another thing to maybe test is an element with a mix of observed and unobserved attributes to ensure the implementation is making the decision about firing attributeChanged on a per-attribute and not per-element basis.

File: custom-elements/reactions/resources/reactions.js [hunk 6]
```diff
@@ -78,6 +78,38 @@ function testNodeDisconnector(testFunction, name) {
...
+        container.appendChild(instance);
+
+        instance.setAttribute('id', 'foo');
+        assert_array_equals(element.log().types(), ['constructed', 'connected', 'attributeChanged']);
+        var newInstance = testFunction(instance);
+        var logEntries = element.log();
+        assert_array_equals(logEntries.types(), ['constructed', 'attributeChanged']);
+        assert_equals(logEntries.log().name, 'id');
+        assert_equals(logEntries.log().oldValue, null);
+        assert_equals(logEntries.log().newValue, 'foo');
```
Comment (dominiccooney) [inline custom-elements/reactions/resources/reactions.js]: Maybe test the namespace parameter here, too?

File: custom-elements/reactions/DOMTokenList.html [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+<script>
+
+test(function () {
+    var element = defineNewCustomElement(['class']);
+    var instance = document.createElement(element.name);
+    assert_array_equals(element.log().types(), ['constructed']);
+    instance.classList.add('foo');
+    var logEntries = element.log();
+    assert_array_equals(logEntries.types(), ['attributeChanged']);
+    assert_equals(logEntries.log().name, 'class');
```
Comment (dominiccooney) [inline custom-elements/reactions/DOMTokenList.html]: This pattern of four assertions is repetitive; I wonder if the tests would be more readable if you add a helper which takes logEntries.log() and four parameters and have the helper do the four assertions?

File: custom-elements/reactions/resources/reactions.js [hunk 8]
```diff
@@ -78,6 +78,38 @@ function testNodeDisconnector(testFunction, name) {
...
+    test(function () {
+        var element = defineNewCustomElement(['id']);
+        var instance = document.createElement(element.name);
+        container.appendChild(instance);
+
+        instance.setAttribute('lang', 'en');
+        assert_array_equals(element.log().types(), ['constructed', 'connected']);
+        var newInstance = testFunction(instance);
+        assert_array_equals(element.log().types(), ['constructed']);
+    }, name + ' must not enqueue a attributeChanged reaction when cloning an element with an unobserved attribute');
```
Comment (dominiccooney) [inline custom-elements/reactions/resources/reactions.js]: Is there a way to make this "an element with an unobserved attribute" more accurate but equally succinct? It must not queue an attribute changed callback for *unobserved attributes*; but a custom element must queue an attributeChanged callback for an observed attribute, even if there happen to be other unobserved attributes. Same goes for the earlier file.

**Review by dominiccooney — COMMENTED** [review]
Thorough tests! Comments inline.

File: custom-elements/reactions/DOMTokenList.html (hunk 1 above)
Comment (rniwa) [inline custom-elements/reactions/DOMTokenList.html]: Removed

File: custom-elements/reactions/DOMTokenList.html (hunk 7 above)
Comment (rniwa) [inline custom-elements/reactions/DOMTokenList.html]: Alright, added assert_attribute_log_entry to reactions.js.

File: custom-elements/reactions/DOMTokenList.html (hunk 2 above)
Comment (rniwa) [inline custom-elements/reactions/DOMTokenList.html]: Oops, fixed that.

File: custom-elements/reactions/DOMTokenList.html (hunk 3 above)
Comment (rniwa) [inline custom-elements/reactions/DOMTokenList.html]: Oops, fixed

File: custom-elements/reactions/DOMTokenList.html (hunk 4 above)
Comment (rniwa) [inline custom-elements/reactions/DOMTokenList.html]: Fixed

File: custom-elements/reactions/DOMTokenList.html (hunk 5 above)
Comment (rniwa) [inline custom-elements/reactions/DOMTokenList.html]: Okay, I think we can address that by using non-empty attribute list.

Comment (rniwa) [inline custom-elements/reactions/DOMTokenList.html]: And added a test case for adding multiple tokens and removing multiple tokens.

File: custom-elements/reactions/resources/reactions.js (hunk 8 above)
Comment (rniwa) [inline custom-elements/reactions/resources/reactions.js]: Sure, I added a new test case. But I think in general, those edge cases are better tested in https://github.com/w3c/web-platform-tests/pull/3853 which targets each reaction type more throughly.  These individual CEReaction tests are just sanity check for each IDL attribute so I don't think we necessary have to exhaustively test each edge case.

File: custom-elements/reactions/resources/reactions.js (hunk 6 above)
Comment (rniwa) [inline custom-elements/reactions/resources/reactions.js]: Sure, now that I'm using a helper, that's done.

**Review by dominiccooney — COMMENTED** [review]
LGTM

**Pull request merged**